doc/specs/stdlib_stringlist.md

```diff
@@ -0,0 +1,467 @@
+---
+title: stringlist
+---
+# Lists of strings
+
+[TOC]
+
+## Introduction
+
+Fortran has supported variable-length strings since the 2003 standard,
+but it does not have a native type to handle collections of strings of
+different lengths. Such collections are quite useful though and the
+language allows us to define a derived type that can handle such
+collections.
+
+The `stdlib_stringlist` module defines a derived type that is capable of
+storing a list of strings and of manipulating them.
+
+Methods include:
+
+* inserting strings at a given position
+* replacing strings at a given position
+* deleting a single string or a range of strings
+* retrieving a string or a range of strings at a given position
+* finding the position of a particular string or a string which contains some substring
+* sorting the list
+
+## Positions in a list of strings
+
+The module implements what are effectively infinitely long lists: a position is
+represented as a positive integer, but there is no "out-of-bound" index. That is,
+the following piece of code will simply work:
+
+```fortran
+type(stringlist_type) :: list
+
+! Add two strings ...
+call list%insert( list_head, "The first string" )
+call list%insert( 20, "The last string" )
+
+write(*,*) 'The last:    ', list%get(list_end)
+write(*,*) 'Beyond that: ', list%get(30)
+```
+The special position `list_head` represents the position *before* the first element. Likewise,
+the special position `list_end` represents the position of the *last* element. You can
```

Comment:
This feels inconsistent to me. `list_head` represents an element outside of the list, while `list_end` represents an element in the list. I would prefer either having both special indices refer to elements inside or outside the list, but not mixed.

Reply:
One surprising consequence of the current choice of `list_head` (`0`) and `list_end` (`-1`) is that appending with `insert` is impossible:
```f90
call list%insert(list_end + 1, "last item")
```
Because it will evaluate to `list_head` and prepend instead.

Reply:
One surprising consequence of the current choice of `list_head` (`0`) and `list_end` (`-1`) is that appending with `insert` is impossible:
```f90
call list%insert(list_end + 1, "last item")
```
Because it will evaluate to `list_head` and prepend instead.

Reply:
Yes, there is that ... I hope to solve this with special types rather than special values.